src/player.c

```diff
@@ -419,28 +435,47 @@ static int play (player_t * const player) {
 					break;
 				}
 
-				const int numChannels = av_get_channel_layout_nb_channels (
-						filteredFrame->channel_layout);
-				const int bps = av_get_bytes_per_sample(filteredFrame->format);
-				ao_play (player->aoDev, (char *) filteredFrame->data[0],
-						filteredFrame->nb_samples * numChannels * bps);
-
-				av_frame_unref (filteredFrame);
+				/*Insert frame for playing*/
+				pthread_mutex_lock (&player->aoplay_lock);
+				STAILQ_INSERT_TAIL (&player->queue_head, frm_element, entry);
```

Comment:
This loop can run multiple times (in theory; the filter graph we’re using does not duplicate frames and we’re queueing only one frame at a time), but you’re always queueing the same frm_element. Thus this is not correct.

Also also: I’m wondering if we could make the cut between decode/playback here, i.e. buffersrc_write_frame is decoding and buffersink_get_frame is playback. That way we wouldn’t need to create our own queue. buffersrc buffers decoded frames (before filter graph) in a fifo, so this might actually work.

Reply:
Good point, let's create `buffer.c` and `buffer.h`